[DDP] Use NCCL allocated memory for gradient bucket

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #146589

So that NVLink SHARP comes with zero-copy on H100+ platforms, for DDP applications.
Less SM usage, less memory contention between NCCL kernel and compute kernels.

Added env `DDP_DISABLE_COMM_MEM` as a back-out option:
```
An environment variable to disable comm-optimized memory pool.
Default is 0, which means comm-optimized memory pool is enabled.
Users can set it to 1 in case of seeing regression or OOM (because this
comm MemPool may not share space with regular compute MemPool).
```

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

Differential Revision: [D69297766](https://our.internmc.facebook.com/intern/diff/D69297766)

kwen2501 committed:
[DDP] Use NCCL allocated memory for gradient bucket

[ghstack-poisoned]

kwen2501 committed:
Update on "[DDP] Use NCCL allocated memory for gradient bucket"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

Skylion007 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -5423,6 +5425,45 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
+        reinterpret_cast<c10::cuda::CUDACachingAllocator::CUDAAllocator*>(
+            getMemAllocator().get());
+    // Pool is created
+    memPool_ = new c10::cuda::MemPool(allocator);
```
Should we explicitly use `new` and delete or would handling it with a unique_ptr be better?

kwen2501 replied:
Yeah, that's a good idea. I used `unique_ptr` originally but realized that almost all APIs around `MemPool` takes raw pointers -- so I had to use a lot of `.get()`. Any recommendation for this situation if we use `unique_ptr`?

Skylion007 replied:
@kwen2501 Annoying for the get() calls, but still better than manual memory management for a whole host of reasns.

kwen2501 replied:
Thanks!

syed-ahmed replied:
`unique_ptr` looks good to me because ProcessGroupNCCL has ownership of the pool.

Skylion007 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -1294,6 +1301,9 @@ class TORCH_API ProcessGroupNCCL : public Backend {
   std::optional<bool> useNonblocking_{std::nullopt};
+
+  // Communication-optimized memory pool associated with this PG
+  c10::cuda::MemPool* memPool_ = nullptr;
```
Why not have it be a std::unique_ptr?

Skylion007 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1543,6 +1543,8 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
               << "ProcessGroupNCCL onCompletionHookThread thread joined.";
   }
+
+  delete memPool_;
```
Use a uniquePtr remove error prone deletes

Skylion007 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -5423,6 +5425,45 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
+    long size,
+    at::TensorOptions options) {
+  // Some checks
+  TORCH_CHECK(options.has_device(), "Tensor options must include device");
```
```suggestion
  TORCH_CHECK_VALUE(options.has_device(), "Tensor options must include device");
```

Skylion007 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -5423,6 +5425,45 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
+  // Some checks
+  TORCH_CHECK(options.has_device(), "Tensor options must include device");
+  auto device = options.device();
+  TORCH_CHECK(
```
```suggestion
  TORCH_CHECK_VALUE(
```

kwen2501 committed:
Update on "[DDP] Use NCCL allocated memory for gradient bucket"


So that NVLink SHARP is enabled on H100+ platforms.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

syed-ahmed commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -5423,6 +5423,45 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
+    // Pool is created
+    memPool_ = std::make_unique<c10::cuda::MemPool>(allocator);
+    // Also need to ncclCommRegister it so that NCCL recognizes it
+    registerMemPool(memPool_.get());
```
This call should happen *after* the tensor is allocated. Inside `registerMemPool`, we call snapshot, which returns all the blocks held by the pool. The pointers of those blocks are then registered with NCCL via ncclCommRegister. So right now, this is registering nothing.

syed-ahmed commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -5423,6 +5423,45 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
+    LOG(INFO) << logPrefix() << "Created memory pool";
+  }
+
+  // Allocate tensor under this MemPool's context
```
Add a `deregisterMemPool` call here, otherwise we'll hit this `TORCH_CHECK`: https://github.com/pytorch/pytorch/blob/33191bb664fef338d94586a05bf1f14d17a00340/torch/csrc/distributed/c10d/NCCLUtils.hpp#L477-L482

syed-ahmed commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -5423,6 +5423,45 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
+  // Allocate tensor under this MemPool's context
+  auto ctx = c10::cuda::MemPoolContext(memPool_.get());
+  c10::cuda::CUDACachingAllocator::beginAllocateToPool(
+      memPool_->device(), memPool_->id(), [](cudaStream_t) { return true; });
```
Add a `deregisterMemPool` call here, otherwise we'll hit this `TORCH_CHECK`: https://github.com/pytorch/pytorch/blob/33191bb664fef338d94586a05bf1f14d17a00340/torch/csrc/distributed/c10d/NCCLUtils.hpp#L477-L482

syed-ahmed commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -5423,6 +5423,45 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
+  auto ctx = c10::cuda::MemPoolContext(memPool_.get());
+  c10::cuda::CUDACachingAllocator::beginAllocateToPool(
+      memPool_->device(), memPool_->id(), [](cudaStream_t) { return true; });
+  at::Tensor tensor = at::empty({size}, options);
```
Add the `registerMemPool(memPool_.get());` here. After that, all pointers in the pool will get registered.

syed-ahmed reviewed (changes requested):
Thanks for the PR! Left a couple comments on registration usage. basically, 
```
deregisterMemPool(...)
at::Tensor tensor = ...
registerMemPool(...)
```

Otherwise, LGTM!

kwen2501 commented:
Thanks @syed-ahmed for the comments! I totally missed the segment level in the MemPool! I wonder if there is a way to get incremental segments? (so that we just register the new ones)
Maybe diff'ing two snapshots?

kwen2501 commented:
Perhaps we can maintain a segment-address-to-bool dictionary to record the registration status of segments?

yifuwang commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -1937,7 +1937,8 @@ def first_bucket_size(ddp_bucket_mb):
                         has_half = any(p.dtype is torch.half for p in m.parameters())
-                        tol = 1.0e-3 if has_half else 1.0e-5
+                        atol = 1.0e-3 if has_half else 1.0e-4
+                        rtol = 1.0e-3
```
Hmm, reduction over registered buffer should lead to better precision if using NVLS, and at least should not be worse than the non-registered baseline if using ring. Do you know why this needs to be relaxed?

kwen2501 replied:
In general yes, but I am not 100% sure about a two-rank case (if that's the world size of this test)

yifuwang replied:
Hmm, IIUC, with rank=2, floating point addition shouldn't suffer from associativity problem at all. Which means we should expect identical results?

syed-ahmed commented:
> Perhaps we can maintain a segment-address-to-bool dictionary to record the registration status of segments? (Edit: I see that registeredSegmentHandles_ already does so)

Yeah, may be we can make `registerSegment` a no-op when a segment is already registered. However, that function was originally implemented for the `TORCH_NCCL_USE_TENSOR_REGISTER_ALLOCATOR_HOOK` path: https://github.com/pytorch/pytorch/blob/3f5ed0568886ca4c144213a27918019a2e3ebab0/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L311-L328. So not sure if it would break any internal workload.

kwen2501 commented:
Yeah, I am adding a flag `errorOnRereg` to `registerSegment` API with a default value of true. In `registerMemPool`, we can set the flag to false.

kwen2501 committed:
Update on "[DDP] Use NCCL allocated memory for gradient bucket"


So that NVLink SHARP is enabled on H100+ platforms for DDP applications.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

syed-ahmed commented:
We should also add a `deregisterMemPool` in ProcessGroupNCCL::shutdown(), and I think that will make this diff complete :).

kwen2501 commented:
I left `deregisterMemPool` out because I wasn't sure when it would be safe to call it --
- If we call it before `ncclCommFinalize`, at that point the collectives are still executing, would the dereg create any side effect? (not very likely but I cannot rule it out)
- If we call it after `ncclCommFinalize`, at that point could some resources of the communicator be freed already and cause the dereg call to fail? 
I'd need more confirmation before adding in that call.

fegin commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -1157,14 +1157,37 @@ void Reducer::initialize_buckets(
+        // Comm-optimized memory pool is available, use it to allocate tensor
+        LOG(INFO)
+            << "Reducer: found comm-optimized memory allocator, using it to create bucket";
+        bucket.gradients = backend->allocateTensor(bucketSize, options);
```
noob question, is the available NCCL available memory size the same? If it is smaller, can it lead to easier to OOM?

kwen2501 replied:
The NCCL memory pool does not have a limited size per my understanding. It is extendable as like regular CUDA caching allocator.

syed-ahmed commented:
Looking at NCCL source, it looks like `ncclCommDeregister` is just cleaning up some cache. Also `ncclCommDestroy` calls the cleanup function as well. I think it's safe to put it after `ncclCommFinalize` as `ncclCommDestroy` happens after `ncclCommFinalize`.
```
ncclCommDeregister -> commDeregister -> regCleanup
ncclCommDestroy -> commReclaim -> commCleanup -> commFree -> ncclRegCleanup -> regCleanup

```

kwen2501 committed:
Update on "[DDP] Use NCCL allocated memory for gradient bucket"


So that NVLink SHARP comes with zero-copy on H100+ platforms, for DDP applications.
Less SM usage, less memory contention between NCCL kernel and compute kernels.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@kwen2501 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69297766).

kwen2501 committed:
Update on "[DDP] Use NCCL allocated memory for gradient bucket"


So that NVLink SHARP comes with zero-copy on H100+ platforms, for DDP applications.
Less SM usage, less memory contention between NCCL kernel and compute kernels.

Added env `DDP_DISABLE_COMM_MEM` as a back-out option:
```
An environment variable to disable comm-optimized memory pool.
Default is 0, which means comm-optimized memory pool is enabled.
Users can set it to 1 in case of seeing regression or OOM (because this
comm MemPool may not share space with regular compute MemPool).
```

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

Differential Revision: [D69297766](https://our.internmc.facebook.com/intern/diff/D69297766)

[ghstack-poisoned]

kwen2501 commented:
@kwen2501 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69297766).

c-p-i-o commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -1157,14 +1157,44 @@ void Reducer::initialize_buckets(
+        LOG(INFO)
+            << "Reducer: default comm backend not found, skipping bucket memory optimization";
+      }
+      if (ddpDisableCommMem == 0 && backend != nullptr &&
```
Do we like `ddpDisableCommMem == 0` or `!ddpDisableCommMem` (because the var is a bool) or it doesn't matter?
just a clarifying question - if in pytorch do we have a style guide for this?

c-p-i-o commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -5423,6 +5432,46 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
 }
 
+at::Tensor ProcessGroupNCCL::allocateTensor(
+    long size,
```
is there a max/recommended size limit?

c-p-i-o commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -1157,14 +1157,44 @@ void Reducer::initialize_buckets(
+      // Users can set it to 1 in case of seeing regression or OOM (because this
+      // comm MemPool may not share space with regular compute MemPool).
+      bool ddpDisableCommMem =
+          (getCvarString({"DDP_DISABLE_COMM_MEM"}, "0") == "1");
```
do you think it's worth logging the value of this env variable in `ProcessGroupNCCL` so we can search for it in logs?

fduwjj replied:
Or ddp log since this is part of DDP.

fduwjj commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -1157,14 +1157,44 @@ void Reducer::initialize_buckets(
+      auto bucketSize = static_cast<long>(offset);
+      // Check if we can use comm-optimized memory pool to allocate tensor
+      c10::intrusive_ptr<Backend> backend = nullptr;
+      // An environment variable to disable comm-optimized memory pool.
```
It would be great that we can explain in the comment why or what benefit of doing this. What you wrote in the summary of this PR is a good fit for it.

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -5423,6 +5432,46 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
+  // Also need to ncclCommRegister the pool in case new segments are created;
+  // reregistration of old segments will be ignored
+  registerMemPool(memPool_.get());
+  c10::cuda::CUDACachingAllocator::endAllocateToPool(
+      memPool_->device(), memPool_->id());
```
Will this cause fragmentation?

fduwjj reviewed (approved):
unblock and in general looks good. I have one more suggestion for Meta internal use case which will follow up with you offline.

kwen2501 commented:
@pytorchbot merge

kwen2501 commented:
@pytorchbot merge

kwen2501 commented:
@pytorchbot merge -f "Don't know why regular merge timed out twice; force merging now"

Skylion007 commented:
Do we need to do anything, should we do anything for other NCCL comms like all_gather, all_scatter, etc?
